kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -304,20 +686,72 @@ impl<'tcx> GotocCtx<'tcx> {
-                self.codegen_ty(res_ty),
-                operands
-                    .iter()
-                    .filter_map(|o| {
-                        let oty = self.operand_ty(o);
-                        if oty.is_unit() { None } else { Some(self.codegen_operand(o)) }
-                    })
-                    .collect(),
-                &self.symbol_table,
-            ),
-            AggregateKind::Adt(_, _, _, _, _) => unimplemented!(),
-            AggregateKind::Closure(_, _) => unimplemented!(),
-            AggregateKind::Generator(_, _, _) => unimplemented!(),
+            AggregateKind::Tuple => {
+                let layout = self.layout_of(res_ty);
+                Expr::struct_expr_from_values(
+                    self.codegen_ty(res_ty),
+                    layout
+                        .fields
+                        .index_by_increasing_offset()
+                        .map(|idx| self.codegen_operand(&operands[idx]))
+                        .collect(),
+                    &self.symbol_table,
+                )
+            }
+            AggregateKind::Adt(_, _, _, _, Some(active_field_index)) => {
+                assert!(res_ty.is_union());
+                assert_eq!(operands.len(), 1);
+                let typ = self.codegen_ty(res_ty);
+                let components = typ.lookup_components(&self.symbol_table).unwrap();
+                Expr::union_expr(
+                    typ,
+                    components[active_field_index].name(),
+                    self.codegen_operand(&operands[0]),
+                    &self.symbol_table,
+                )
+            }
+            AggregateKind::Adt(_, _, _, _, _) if res_ty.is_simd() => {
+                let typ = self.codegen_ty(res_ty);
+                let layout = self.layout_of(res_ty);
+                Expr::vector_expr(
+                    typ,
+                    layout
+                        .fields
+                        .index_by_increasing_offset()
+                        .map(|idx| self.codegen_operand(&operands[idx]))
+                        .collect(),
+                )
+            }
+            AggregateKind::Adt(_, variant_index, _, _, _) if res_ty.is_enum() => {
+                self.codegen_rvalue_enum(&variant_index, &operands, res_ty)
+            }
+            AggregateKind::Adt(_, _, _, _, _) => {
+                let typ = self.codegen_ty(res_ty);
+                let layout = self.layout_of(res_ty);
+                Expr::struct_expr_from_values(
+                    typ,
+                    layout
+                        .fields
+                        .index_by_increasing_offset()
+                        .map(|idx| self.codegen_operand(&operands[idx]))
+                        .collect(),
+                    &self.symbol_table,
+                )
+            }
+            AggregateKind::Closure(_, _) => {
+                let typ = self.codegen_ty(res_ty);
```

Comment:
This logic is replicated in at least three different places. You could just join the match statements, like:
```rust
AggregateKind::Tuple | AggregateKind::Closure(..) | AggregateKind::Adt(..) => {
   // code
}
```

Reply:
Thanks, done.